apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -0,0 +1,180 @@
+            GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+
+            clientTask = clientTask ?? TraceServiceClient.CreateAsync();
+            config = config ?? TraceConfiguration.Create();
+
+            IConsumer<TraceProto> consumer = ConsumerFactory<TraceProto>.GetConsumer(
+                 new GrpcTraceConsumer(clientTask), TraceSizer.Instance, config.BufferOptions);
+
+            services.AddSingleton(new ProjectId { Id = projectId });
+            services.AddSingleton(consumer);
+            services.AddSingleton(TraceIdFactory.Create());
+            services.AddSingleton(RateLimitingTraceOptionsFactory.Create(config));
+            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+
+            services.AddScoped(CreateTraceHeaderContext);
+            services.AddScoped(CreateManagedTracer);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// </summary>
+        internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
+        {
+            var accessor = provider.GetService<IHttpContextAccessor>();
+            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            return TraceHeaderContext.FromHeader(header);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IManagedTracer"/> based on the <see cref="TraceHeaderContext"/> and
+        /// the rate limiter.
+        /// </summary>
+        internal static IManagedTracer CreateManagedTracer(IServiceProvider provider)
+        {
+            var headerContext = provider.GetService<TraceHeaderContext>();
+            var rateLimitingFactory = provider.GetService<RateLimitingTraceOptionsFactory>();
+            var projectId = provider.GetService<ProjectId>();
+            var traceIdFactory = provider.GetService<TraceIdFactory>();
+            var consumer = provider.GetService<IConsumer<TraceProto>>();
+
+            var message = "No {0} service found. Ensure Google Cloud Trace is properly set up.";
+            GaxPreconditions.CheckState(headerContext != null, 
+                string.Format(message, typeof(TraceHeaderContext).GetType()));
+            GaxPreconditions.CheckState(rateLimitingFactory != null,
+                string.Format(message, typeof(RateLimitingTraceOptionsFactory).GetType()));
+            GaxPreconditions.CheckState(projectId != null, 
+                string.Format(message, typeof(ProjectId).GetType()));
+            GaxPreconditions.CheckState(traceIdFactory != null, 
+                string.Format(message, typeof(TraceIdFactory).GetType()));
+            GaxPreconditions.CheckState(consumer != null, 
+                string.Format(message, typeof(TraceProto).GetType()));
+
+            // If the trace header and rate limiter say not to trace, return a no-op tracer.
```

Comment:
Yea, I'm ignoring this case actually.  I will do this in a follow up PR as I need to change some common files and the not core version